Player and dealer

nwarsalee opened the pull request:
Implemented the Player and the Dealer classes for the Blackjack simulator.

Both have a compute_play() function that is their main use.
This function will return an enum representing their decision based on their blackjack hand.

Both classes include additional functions and members to aid in the calculation of the decision.

- [x] @esulu 
- [ ] @cailynse

nwarsalee committed:
Create Dealer and Player files

Created Dealer and Player files.

Added initial helper functions to the Player class to help with reading and using their hand.

nwarsalee committed:
Implement Decision Making for Player and Dealer

Implemented the decision logic that is to be used by both the player and the dealer.

Player utilizes Basic Strategy rules.
Dealer utilizes dealer's default strategy of stand on 17 above.

Added more helper functions to Dealer, that are similar if not the same as the ones in Player.

Now both Player and Dealer make use of enum defined in definitions.py.

nwarsalee committed:
Add Print of Person hand and decision

Added a print function in both the Player and Dealer class that will print their given hand and the decision that was determined based on their respective strategies.

Added an attribute to both Dealer and Player called dev_mode.
Default value for this is set to False.
When set to True, it will activate all print and debug statements for the class (currently only activates the print_decision function)

nwarsalee committed:
Change Ace point addition from 10 to 11

Fixed problem where it was adding 10 onto the hand total instead of 11 for an Ace.

esulu commented on dealer.py:
```diff
@@ -0,0 +1,177 @@
+            curr_min = min(totals)
+
+            # error case
+            if curr_max > 21 and curr_min > 21:
```
Aren't curr_max and curr_min still lists here? Do we mean to compare their sums? (Same applies with the player class). 

Also I think the lists are being returned rather than the integer sum value

esulu replied:
nvm python list comparison op

esulu replied:
But I believe a list is being returned rather than an int still

cailyn-codes replied:
max returns the largest item in the given array, so it will be the highest/lowest in the array of totals.

esulu replied:
Oh I thought total was holding the lists of the hands, looks good!

esulu commented on dealer.py:
```diff
@@ -0,0 +1,177 @@
+
+        return 1 in hand
+
+    def is_pair(hand, card_type):
```
Is this necessary for the dealer?

nwarsalee replied:
No don't believe so, just ported it from the Dealer class.
Will remove.

cailyn-codes commented on dealer.py:
```diff
@@ -0,0 +1,177 @@
+        decision = definitions.Actions.STAND
+
+        # Var to store hand total
+        total = Dealer.valid_total(Dealer.hand_total(hand))
```
I think this value (`Dealer.hand_total(hand)` will occasionally be an int and not a list? (see comment above)

nwarsalee replied:
In the docstring for the function I mistakenly put it returns an int instead of list of ints.
In the previous comment, I mention how the function should always return a list.

cailyn-codes commented on dealer.py:
```diff
@@ -0,0 +1,177 @@
+                    total[0] += val
+                    total[1] += val
+            
+            return total
```
I think this should be `list(total)` - or we need to update `valid_total`

nwarsalee replied:
At line 78 I make the _total_ list with all initial values set to 0.
So when returning total, it should be in the form of a list.

cailyn-codes reviewed (commented):
Looks good! I'd love to see some cursory tests or sample test output.

nwarsalee committed:
Remove is_pair function from Dealer

Removed is_pair function from the Dealer class.

Doesn't ever seem to be used.

nwarsalee committed:
Update hand_total comment

nwarsalee committed:
Rework Basic Strategy and Use PyDealer

Reworked basic strategy logic that player uses  to be in line with what is in the README for the project.

Updated both the Player and the Dealer class to make use of the PyDealer module for cards.

Both now expect a list of Card objects when computing their decision.

Helper functions have been modified to conform with PyDealer's API and created new helper functions where needed.

esulu commented on player.py:
```diff
@@ -169,85 +202,129 @@ def compute_play(hand, dealer_upcard):
+                if dealer_up_val in range(4, 7):
+                    decision = definitions.Actions.DOUBLE
+                else:
+                    decision = definitions.Actions.STAND
```
this should be a hit :)

nwarsalee replied:
Woops :), good catch

nwarsalee committed:
Fix Soft Hit case

Fixed error that had player standing when they should be hitting.

nwarsalee committed:
Add Blackjack enum and case for blackjack

Added a new blackjack enum for the Actions enum.

Added another check in the player and dealer's compute strategy to check for a blackjack.

esulu commented:
upstream was in the wrong place.. accidentally merged..
